### PR TITLE
Warning fixes for quaternion headers

### PR DESCRIPTION
**Pull request by savuor**
Backport to 4.x from #21018 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/core/include/opencv2/core/dualquaternion.inl.hpp [hunk 1]
```diff
@@ -65,7 +65,7 @@ DualQuat<T> DualQuat<T>::createFromAngleAxisTrans(const T angle, const Vec<T, 3>
 {
     Quat<T> r = Quat<T>::createFromAngleAxis(angle, axis);
     Quat<T> t{0, trans[0], trans[1], trans[2]};
-    return createFromQuat(r, t * r / 2);
+    return createFromQuat(r, t * r / T(2.0));
```
Comment (alalek) [inline modules/core/include/opencv2/core/dualquaternion.inl.hpp:68]: Lets use `* 0.5f`

File: modules/core/include/opencv2/core/dualquaternion.inl.hpp [hunk 2]
```diff
@@ -158,7 +158,7 @@ inline Quat<T> DualQuat<T>::getRotation(QuatAssumeType assumeUnit) const
 template <typename T>
 inline Vec<T, 3> DualQuat<T>::getTranslation(QuatAssumeType assumeUnit) const
 {
-    Quat<T> trans = 2.0 * (getDualPart() * getRealPart().inv(assumeUnit));
+    Quat<T> trans = T(2.0) * (getDualPart() * getRealPart().inv(assumeUnit));
```
Comment (alalek) [inline modules/core/include/opencv2/core/dualquaternion.inl.hpp:161]: `2.0f` ?

File: modules/core/include/opencv2/core/dualquaternion.inl.hpp (hunk 1 above)
Comment (savuor) [reply modules/core/include/opencv2/core/dualquaternion.inl.hpp]: Makes sense, fixed to `* 0.5`

File: modules/core/include/opencv2/core/dualquaternion.inl.hpp (hunk 2 above)
Comment (savuor) [reply modules/core/include/opencv2/core/dualquaternion.inl.hpp:161]: `T` type can be float or double, so what's the point of using `f` postfix?

Comment (alalek) [reply modules/core/include/opencv2/core/dualquaternion.inl.hpp:161]: f suffix makes number of float type.
AFAIK, there is no issues with both operations `float*float` or `float*double` (the second result is double).

Comment (savuor) [reply modules/core/include/opencv2/core/dualquaternion.inl.hpp:161]: It's understood; the problem is that this is not `float*float` or `float*double`, this is `float*Quat<T>` where `T` can be both.
No wonder using 2.0f breaks compilation on my machine (fails to find corresponding operator).

I propose to leave this code as it is now.

**Review by alalek — APPROVED** [review]

**Pull request merged**